### ISSUE TITLE
_dumbwin32proc.Process does not implement IProcessTransport completely

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2838]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)| @Peaker reported|
|-|-|
|Trac ID|trac#2838|
|Type|defect|
|Created|2007-09-25 16:48:34Z|
|Branch|https://github.com/twisted/twisted/tree/win32-proc-missing-2838-2|

writeToChild(childFD, data) is missing from the implementation of _dumbwin32proc

Attachments:

* [_dumwin32proc.2.py_diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/802/802e3f8209166abf8be5d54be8e47aadc4b20d19/c7def538ee96213cbc80d4338d3d8cb1ea3ab450.py_diff) (1224 bytes) - added by bigdog on 2008-02-11 05:01:37Z - 
* [_dumwin32proc.py_diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/802/802e3f8209166abf8be5d54be8e47aadc4b20d19/2dd7f64ab33928734bbb1187764b34cddf112551.py_diff) (1944 bytes) - added by bigdog on 2008-02-13 05:08:12Z - 
* [test_process.py_diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/802/802e3f8209166abf8be5d54be8e47aadc4b20d19/e7c6deadd9e00d50fd57eca65fcd9ee07a68ec35.py_diff) (3423 bytes) - added by bigdog on 2008-02-13 05:15:45Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2838 2838
type__defect defect
reporter__Peaker Peaker
priority__normal normal
milestone__ 
branch__branches_win32_proc_missing_2838_2 branches/win32-proc-missing-2838-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1190738914000000 1190738914000000
changetime__1317520912000000 1317520912000000
version__None None
owner__exarkun exarkun
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bigdog set owner to bigdog<br>bigdog set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bigdog removed owner<br>bigdog set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to bigdog|
|-|-|

 * the new test should be named test_writeToChild
 * the SkipTest is not necessary, the skip is already made at class level
 * missing spaces after commas and around equal operator
 * assertRaises doesn't take a message argument
 * one thing that would be cool is to use zope.interface.verify.verifyClass in another test to check that the class isn't missing something else

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|bigdog removed owner|
|-|-|

Therve,

Thank you, I am slowly learning the twisted way :-)

Thanks for the tip about verify,  I added a test case to validate that Process implements all functions defined in it's interfaces, and that test identified two additional functions.

getPeer and getHost.

I added those functions, and tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to bigdog|
|-|-|

Please remove the local imports and move the inner classes to a lower level of indentation, plus, explain what you're faking and how much of it you're faking.

It looks like maybe you could fake a lot less stuff if you split the massive `__init__` into (individually tested) pieces; at the very least, an initializer that sets up the appropriate state a "start" method.

It turns out that the UNIX stuff doesn't actually declare that it `implements()` `IProcessTransport` either, so if you could file a ticket for that it would be great, and probably easier to get started with :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [24103]) Branching to 'win32-proc-missing-2838'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [24104]) apply _dumwin32proc.py_diff and test_process.py_diff

refs [#2838](https://github.com/twisted/twisted/issues/2838)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

I'll probably get back to this after [#10883](https://github.com/twisted/twisted/issues/10883) is merged.  Of course, feel free to submit more patches (but make them against the branch instead of against trunk).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [32704]) Branching to 'win32-proc-missing-2838-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

The branch now only adds `writeToChild`, not `getHost` and `getPeer`.  See [#11552](https://github.com/twisted/twisted/issues/11552) for those.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/win32-proc-missing-2838-2)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @exarkun|
|-|-|

Looks pretty good to me.

One thing to consider: on other platforms, `writeToChild` will raise a `KeyError` when an invalid file descriptor is passed.  This error behavior is accidental, and possibly not the best thing that could be done, but I think it would be best to remain consistent with `twisted.internet.process` in this code as much as possible.  Mostly I noticed this because I don't like docstrings that say "only XYZ is supported"; it reminds me of the C standard's "undefined behavior".  If it's an error, we should detect it, and say what kind of error it is.

I can see that `NotImplementedError` might be a clearer thing for the maintainer of code calling this API to see.  In either case, it should say what exception is raised.

Land at your discretion.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Yea, good call, thanks.  I changed it to `KeyError`, updated the interface documentation, and added a test for the exceptional case (r32730).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [32731]) Merge win32-proc-missing-2838-2

Author: bigdog, exarkun
Reviewer: therve, glyph
Fixes: [#2838](https://github.com/twisted/twisted/issues/2838)

Implement `IProcessTransport.writeToChild` on Windows.
